File: api/app/eval.py
"""Modèle d'évaluation : win-probability, classification, précision.

Déterministe, documenté, testable. C'est le fondement "honnête" du projet :
les seuils sont explicites et constants — aucune intervention du LLM ici.

Modèle de probabilité de gain : logistique standard
    wp(side) = 100 / (1 + 10^(-cp/400))
où cp est l'évaluation côté "side" (négatif si désavantage). Les mats sont
assimilés à 0 % ou 100 %.

Classification (style Lichess, simplifiée et documentée) : une perte de
probabilité de gain d'au moins
    20 points      -> blunder
    10 points      -> mistake
    5 points       -> inaccuracy
    2 points       -> good
    sinon          -> best
"""
from __future__ import annotations

from .schemas import EvalPoint

BEST = "best"
GOOD = "good"
INACCURACY = "inaccuracy"
MISTAKE = "mistake"
BLUNDER = "blunder"

CLASSIFICATION_ORDER = [BLUNDER, MISTAKE, INACCURACY, GOOD, BEST]

# notes par classification (pour la précision)
MOVE_SCORE = {BEST: 1.0, GOOD: 0.7, INACCURACY: 0.4, MISTAKE: 0.1, BLUNDER: 0.0}


def win_prob(eval_point: EvalPoint, side: str) -> float:
    """Probabilité de gain (0-100) du côté `side` ('w'/'b')."""
    cp = eval_point.cp
    if eval_point.mate is not None:
        if side == "w":
            return 100.0 if eval_point.mate > 0 else 0.0
        return 100.0 if eval_point.mate < 0 else 0.0
    # cp est toujours exprimé du point de vue Blancs par le moteur.
    if side == "b":
        cp = -cp
    return max(0.0, min(100.0, 100.0 / (1.0 + 10 ** (-cp / 400.0))))


def classify(eval_before: EvalPoint, eval_after: EvalPoint, side: str) -> tuple[str, float, float, float]:
    """Classe le coup joué. Renvoie (classification, wp_avant, wp_après, perte)."""
    wp_before = win_prob(eval_before, side)
    wp_after = win_prob(eval_after, side)
    loss = wp_before - wp_after
    if loss >= 20:
        cls = BLUNDER
    elif loss >= 10:
        cls = MISTAKE
    elif loss >= 5:
        cls = INACCURACY
    elif loss >= 2:
        cls = GOOD
    else:
        cls = BEST
    return cls, wp_before, wp_after, loss


def accuracy(scores: list[float]) -> float:
    """Précision (0-100) = moyenne pondérée des scores de coups du joueur."""
    if not scores:
        return 0.0
    return round(100.0 * sum(scores) / len(scores), 1)


def acpl_from_losses(losses: list[float]) -> float:
    """ACPL : moyenne des pertes de probabilité, bornée à 0 par coup
    (un coup meilleur que l'attente = gain, pas de perte négative)."""
    if not losses:
        return 0.0
    return round(sum(max(0.0, loss) for loss in losses) / len(losses), 1)

File: web/src/lib/shared/eval.ts
/**
 * Portage TypeScript du modèle d'évaluation backend (api/app/eval.py).
 * Déterministe et identique au serveur : win-probability logistique,
 * classification style Lichess, précision et ACPL.
 */
export interface EvalPoint {
  cp: number | null;
  mate: number | null;
}

export const BEST = "best";
export const GOOD = "good";
export const INACCURACY = "inaccuracy";
export const MISTAKE = "mistake";
export const BLUNDER = "blunder";

export const CLASSIFICATION_ORDER = [BLUNDER, MISTAKE, INACCURACY, GOOD, BEST];

export const MOVE_SCORE: Record<string, number> = {
  [BEST]: 1.0,
  [GOOD]: 0.7,
  [INACCURACY]: 0.4,
  [MISTAKE]: 0.1,
  [BLUNDER]: 0.0,
};

export function hasScore(evalPoint: EvalPoint): boolean {
  return evalPoint.mate !== null || evalPoint.cp !== null;
}

/** Probabilité de gain (0-100) du côté `side` depuis une évaluation UCI
 * (score côté camp au trait `stm`). */
export function winProb(evalPoint: EvalPoint, stm: string, side: string): number {
  if (evalPoint.mate !== null) {
    const stmWins = evalPoint.mate > 0;
    if (side === stm) return stmWins ? 100 : 0;
    return stmWins ? 0 : 100;
  }
  const cp = evalPoint.cp;
  if (cp === null) return 50;
  let cpW = stm === "w" ? cp : -cp;
  if (side === "b") cpW = -cpW;
  return Math.max(0, Math.min(100, 100 / (1 + Math.pow(10, -cpW / 400))));
}

export interface ClassifyResult {
  classification: string;
  wpBefore: number;
  wpAfter: number;
  loss: number;
}

export function classify(
  evalBefore: EvalPoint,
  evalAfter: EvalPoint,
  side: string,
  wpBefore: number | null = null,
  wpAfter: number | null = null,
): ClassifyResult {
  const before = wpBefore !== null ? wpBefore : winProb(evalBefore, side, side);
  const after =
    wpAfter !== null
      ? wpAfter
      : winProb(evalAfter, side === "b" ? "w" : "b", side);
  const loss = before - after;
  let cls: string;
  if (loss >= 20) cls = BLUNDER;
  else if (loss >= 10) cls = MISTAKE;
  else if (loss >= 5) cls = INACCURACY;
  else if (loss >= 2) cls = GOOD;
  else cls = BEST;
  return { classification: cls, wpBefore: before, wpAfter: after, loss };
}

export function accuracy(scores: number[]): number {
  if (!scores.length) return 0;
  return Math.round((100 * scores.reduce((a, b) => a + b, 0)) / scores.length * 10) / 10;
}

export function acplFromLosses(losses: number[]): number {
  if (!losses.length) return 0;
  const avg = losses.reduce((a, b) => a + Math.max(0, b), 0) / losses.length;
  return Math.round(avg * 10) / 10;
}

export const CONCEPT_LABEL_SHARED: Record<string, string> = {
  hanging_piece: "Pièce en prise",
  missed_capture: "Prise manquée",
  fork: "Fourchette",
  pin_moved: "Pièce clouée déplacée",
  pin_missed: "Clouage manqué",
  missed_mate: "Mat manqué",
  candidate: "Coup candidat manqué",
  allowed_mate: "Mat subi",
  back_rank: "Mat de la première rangée",
  roque_missed: "Roque manqué",
  pawn_structure: "Faiblesse de pions",
  promotion: "Promotion manquée",
  passed_pawn: "Pion passé négligé",
  development: "Développement insuffisant",
  threat_ignored: "Menace adverse négligée",
};